contracts/modules/RarityShuffleMetadata.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import { LibString } from "solady/utils/LibString.sol";
+import { ISoundEditionV1a } from "@core/interfaces/ISoundEditionV1a.sol";
+import { IRarityShuffleMetadata } from "@modules/interfaces/IRarityShuffleMetadata.sol";
+
+/**
+ * @title RarityShuffleMetadata
+ * @dev Module for shuffling each tokenID of a Sound protocol mint and bucketing shuffled IDs into rarity tiers
+ * @author @ipatka
+ */
+contract RarityShuffleMetadata is IRarityShuffleMetadata {
+    mapping(uint256 => uint16) public availableIds; /*Mapping to use to track used offsets*/
+    uint16 public availableCount; /*Track the available count to know the id of the current max offset*/
+    mapping(uint256 => uint256) public offsets; /*Store offsets once found*/
+
+    uint256 public nextIndex; /*Store next index to set*/
+
+    uint256[] public ranges; /*Store range sizes for song Ids*/
+    uint256 public nRanges; /*Store number of rarity tiers*/
+
+    address public edition; /*The address of the edition that can trigger metadata updates*/
+
+    /// @notice Constructor sets the shuffle parameters
+    /// @param _edition Address of collection
+    /// @param _availableCount Max number of offsets
+    /// @param _nRanges Number of rarity tiers
+    /// @param _ranges IDs at which to start each tier, must be ordered smallest to largest
+    constructor(
+        address _edition,
+        uint16 _availableCount,
+        uint256 _nRanges,
+        uint256[] memory _ranges
+    ) {
+        edition = _edition;
+        availableCount = _availableCount; /*Set max offsets*/
+        nRanges = _nRanges;
+        uint256 _currentMax;
+        for (uint256 index = 0; index < nRanges; index++) {
+            if (_ranges[index] < _currentMax) revert RangeMustBeOrdered();
+            ranges.push(_ranges[index]); /*Populate the rarity table*/
+            _currentMax = _ranges[index];
+        }
+        emit NewModuleCreated(_edition, address(this));
+    }
+
+    /// @notice Set shuffled token IDs for a quantity of tokens, determiend by edition
+    /// @param quantity Number of tokens to set
+    function triggerMetadata(uint256 quantity) external {
+        if (msg.sender != edition) revert OnlyEditionCanTrigger();
+
+        // Set shuffled ID for each token
+        for (uint256 index = nextIndex; index < nextIndex + quantity; index++) {
+            uint256 pseudorandomness = uint256(keccak256(abi.encodePacked(blockhash(block.number - 1), index)));
```

Comment:
This makes me think the logic in `updatesMintRandomness` should be moved to a pure function on a library contract. wdyt @Vectorized?

@ipatka  Would you be opposed to using that instead of the above?

Reply:
@gigamesh I think we can leave it there.

Reply:
Couldn't `randomness` be passed as an argument?